proxy/proxy.go

```diff
@@ -86,55 +87,65 @@ func (p *Proxy) ProxyHTTP(
 	ruleSpan.End()
+
 	logger := newHTTPLogger(p.log, tr.ConnIndex, req, ruleNum, rule.Service.String())
 	logHTTPRequest(&logger, req)
+
 	if err, applied := p.applyIngressMiddleware(rule, req, w); err != nil {
 		if applied {
 			logRequestError(&logger, err)
-			return nil
 		}
 		return err
 	}
 
+	if _, isStreamBased := rule.Service.(ingress.StreamBasedOriginProxy); isStreamBased || rule.Config.BastionMode {
+		return p.handleStreamBasedService(rule, req, w, tr, &logger)
+	}
+
+	return p.handleHTTPBasedService(rule, req, w, tr, isWebsocket, &logger)
+}
+
+func (p *Proxy) handleStreamBasedService(rule *ingress.Rule, req *http.Request, w connection.ResponseWriter, tr *tracing.TracedHTTPRequest, logger *zerolog.Logger) error {
+	// If in bastion mode, we need to resolve the destination from the request, so service like http_status:404
+	// won't work since it doesn't have EstablishConnection method to resolve the destination
+	if _, ok := rule.Service.(ingress.StreamBasedOriginProxy); !ok && rule.Config.BastionMode {
+		return fmt.Errorf("Unsupported service to stream to in bastion mode: %s", rule.Service)
+	}
+
+	dest, err := getDestFromRule(rule, req)
+	if err != nil {
+		return err
+	}
+
+	flusher, ok := w.(http.Flusher)
+	if !ok {
+		return fmt.Errorf("response writer is not a flusher")
+	}
+
+	rws := connection.NewHTTPResponseReadWriterAcker(w, flusher, req)
+	if err := p.proxyStream(tr.ToTracedContext(), rws, dest, rule.Service.(ingress.StreamBasedOriginProxy), logger); err != nil {
+		logRequestError(logger, err)
+		return err
+	}
+	return nil
+}
+
+func (p *Proxy) handleHTTPBasedService(rule *ingress.Rule, req *http.Request, w connection.ResponseWriter, tr *tracing.TracedHTTPRequest, isWebsocket bool, logger *zerolog.Logger) error {
 	switch originProxy := rule.Service.(type) {
 	case ingress.HTTPOriginProxy:
-		if err := p.proxyHTTPRequest(
-			w,
-			tr,
-			originProxy,
-			isWebsocket,
-			rule.Config.DisableChunkedEncoding,
-			&logger,
-		); err != nil {
-			logRequestError(&logger, err)
+		if err := p.proxyHTTPRequest(w, tr, originProxy, isWebsocket, rule.Config.DisableChunkedEncoding, logger); err != nil {
+			logRequestError(logger, err)
 			return err
 		}
-		return nil
-	case ingress.StreamBasedOriginProxy:
-		dest, err := getDestFromRule(rule, req)
-		if err != nil {
-			return err
-		}
-		flusher, ok := w.(http.Flusher)
-		if !ok {
-			return fmt.Errorf("response writer is not a flusher")
-		}
-		rws := connection.NewHTTPResponseReadWriterAcker(w, flusher, req)
-		logger := logger.With().Str(logFieldDestAddr, dest).Logger()
-		if err := p.proxyStream(tr.ToTracedContext(), rws, dest, originProxy, &logger); err != nil {
-			logRequestError(&logger, err)
-			return err
-		}
-		return nil
 	case ingress.HTTPLocalProxy:
 		p.proxyLocalRequest(originProxy, w, req, isWebsocket)
-		return nil
 	default:
 		return fmt.Errorf("Unrecognized service: %s, %t", rule.Service, originProxy)
 	}
+	return nil
 }
```

Comment:
extracted the logic into `handleHTTPBasedService` and `handleStreamBasedService`